### ISSUE TITLE
Confusion about PBC in CUDA plugin

### ISSUE THREAD

**Issue by jlmaccal** [pandegroup/openmm#2188]
We're trying to extend our MELD plugin to work correctly with PBCs. What we would like to do is to consider particles connected by MELD restraints to be part of the same molecule, so that they are imaged together as a single unit.

To that end, we have implemented `getBondedParticles` in the `MeldForceImpl`. I can verify that this method is called and is returning the correct information. However, we appear to be getting wrapped coordinates in our cuda kernel, which leads to distances being wrong.

We have the following test system:
- two particles at `(1.0, 0.0, 0.0)` and `(6.0, 0.0, 0.0)`
- box vectors of `(3.0, 0.0, 0.0)`, `(0.0, 3.0, 0.0)`, `(0.0, 0.0, 3.0)`
- we add a `NonbondedForce` with `CutoffPeriodic` to force the system to use PBC
- we add a MELD restraint between the two particles

If I set the positions in the context and then fetch them out in the state, we get `(-2.0, 0.0, 0.0)` and `(3.0, 0.0, 0.0)` as I expect. If I remove the meld force, we get `(1.0, 0.0, 0.0)` and `(0.0, 0.0, 0.0)`, again as expected (because without the MELD force, the two particles are not part of the same molecule).

However, inside of our kernel, `posq` has wrapped coordinates (1.0, 0.0, 0.0) and (0.0, 0.0, 0.0), which isn't what I expect. Basically, I don't understand why I'm getting wrapped coordinates inside of my kernel or how to avoid it.

Also, it isn't clear to me how openmm decides if it should be using PBC or not. My class's `usesPeriodicBoundaryConditions` method is never called.

**Comment by peastman** [comment]
For purposes of reordering atoms within the CUDA and OpenCL platforms, molecules are determined from the groups defined in the `CudaForceInfo` or `OpenCLForceInfo`.  See, for example, https://github.com/pandegroup/openmm/blob/master/platforms/cuda/src/CudaKernels.cpp#L492-L516.  That's what you need to implement to tell it which atoms are identical, which atoms form groups that should be kept together, and which groups are identical (as far as this particular force is concerned).

**Comment by jlmaccal** [comment]
I see! Missed that entirely.

How does this get used? E.g. should I dihedral be a group of 4, or would it be sufficient to give 3 groups of 2, corresponding to the 3 bonds.

Basically, is OpenMM doing enough graph theory to infer that all of those particles need to be images together?

**Comment by peastman** [comment]
A dihedral is one group of 4 particles.  https://github.com/pandegroup/openmm/blob/master/platforms/cuda/src/CudaKernels.cpp#L965-L991

**Comment by jlmaccal** [comment]
Thanks. I see where this is used now. A couple of more questions:

1. As I understand it, the `ForceInfo` object is used for two things: (1) deciding what a molecule is for periodic imaging, and (2) deciding which interactions are identical and can thus be re-indexed. Is this correct?
2. In MELD, `areParticlesIdentical` returns false for any particles involved in a MELD force and `areGroupsIdentical` always returns false. Under these conditions, I think it is sufficient to to only include pairwise links, e.g. i-j, j-k, k-l for a torsion. Does this sound correct?

Basically, I'm trying to re-use the logic from `getBondedParticles`, which returns a vector of pairs.

**Comment by peastman** [comment]
It isn't actually anything to do with imaging or periodic boundary conditions.  That's based on the bond definitions in the ForceImpl.  The CudaForceInfo is used for reordering particles to make the nonbonded forces more efficient.  We don't want to reorder them in any way that would affect the force field definition, since that would require rebuilding data structures.  So it only does reorderings that don't affect force field parameters, exceptions, constraints, etc.  For example, exchanging two identical molecules.

If `areParticlesIdentical()` returns false if either particle is involved in a MELD force, I don't think you even need to worry about groups.  That's already enough to guarantee it won't change the index of any particle MELD is looking at.

**Comment by jlmaccal** [comment]
> It isn't actually anything to do with imaging or periodic boundary conditions. That's based on the bond definitions in the ForceImpl.

Obviously you would know this better than me, but this doesn't seem to be the case. When I only implement `getBondedParticles` on the `ForceImpl`, I run into the imaging problem described in the original question. The coordinates in the `State` are imaged correctly, but the ones that my kernel is receiving are not. If I include all of the bonds as groups in the `CudaForceInfo` and return `false` from both `areParticlesIdentical` and `areGroupsIdentical`, then things seem to work correctly. Basically, it doesn't seem like implementing `getBondedParticles` alone is enough.

It think the current version of my code is working correct, as it passes all of our tests. I'm mostly just trying to clarify this for google-ability when the next person is trying to implement a custom force.

**Comment by frank-stonybrook** [comment]
I have a silly question regarding to displacement calculation between two particles under PBC. Following Justin's first post:

> - two particles at (1.0, 0.0, 0.0) and (6.0, 0.0, 0.0)
> - box vectors of (3.0, 0.0, 0.0), (0.0, 3.0, 0.0), (0.0, 0.0, 3.0)
> - we add a NonbondedForce with CutoffPeriodic to force the system to use PBC
> - we add a MELD restraint between the two particles

Under this system setup, shouldn't the `State`  as well as the `posq ` in MELD kernel give the following positions for two particles: `(1.0, 0.0, 0.0)` and `(0.0, 0.0, 0.0)`. Because the 'true' displacement of these separate particles should be 1.0Å under PBC, right?

**Comment by peastman** [comment]
See https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#periodic.

**Comment by frank-stonybrook** [comment]
Thanks for pointing that out! Let me clarify the context a little bit. There are two potential situation which might involved this MELD restraints:

- MELD restraints are added within a molecule. For example, trying to fold a single protein chain.

- MELD restraints are added between two molecule. For example, trying to bind a ligand to a molecule.

If it were the 1st case, where the two particles are within a same molecule (a.k.a a protein chain), then getting the answer of 
`(-2.0, 0.0, 0.0)` and` (3.0, 0.0, 0.0)` is the expected behavior. 
However, if the situation is the 2nd case, then the expected behavior is to get `(1.0, 0.0, 0.0)` and `(0.0, 0.0, 0.0)`, right?

**Comment by peastman** [comment]
In the second case, neither one is more "expected" than the other.  (3, 0, 0) and (0, 0, 0) each represent the same infinite grid of particles.  Since you're dealing with nonbonded interactions, it doesn't matter which one is specified.  All displacements will be identical either way.

**Comment by frank-stonybrook** [comment]
Thanks! Is there an easy to way to print out those displacement from the context to verify that? I got weird results from a MELD PBC test which made me suspect that the displacements aren't the same.

**Comment by jlmaccal** [comment]
@frank-stonybrook Can we move this to the meld repo? @peastman didn't write MELD and can't be expected to support it. More specifics other than that your results were weird would be helpful.

**Comment by peastman** [comment]
Here's the code used by the reference platform to compute displacements.

https://github.com/openmm/openmm/blob/eec9cd69a6f2c1e92d365691bbc62eae621b481c/platforms/reference/src/SimTKReference/ReferenceForce.cpp#L90-L101

**Comment by frank-stonybrook** [comment]
Thank you so much @peastman! It always super helpful discussing with you. Since my question is related to some details of MELD simulation, I will move the discussion to the MELD repo.
